railties/lib/rails/generators/app_base.rb

```diff
@@ -410,10 +413,55 @@ def os_supports_listen_out_of_the_box?
         RbConfig::CONFIG["host_os"] =~ /darwin|linux/
       end
 
+      def package_json_exist?
+        File.exist?("package.json")
+      end
+
       def run_bundle
         bundle_command("install") if bundle_install?
       end
 
+      def yarn_path
+        commands = ["yarn"]
+
+        if RbConfig::CONFIG["host_os"] =~ /mswin|cygwin/
+          ENV["PATHEXT"].split(File::PATH_SEPARATOR).each do |ext|
+            commands << commands[0] + ext
+          end
+        end
+
+        yarn_path = commands.find do |cmd|
+          paths = ENV["PATH"].split(File::PATH_SEPARATOR)
+          path = paths.find do |p|
+            full_path = File.expand_path(cmd, p)
+            File.executable?(full_path) && File.file?(full_path)
+          end
+          path && File.expand_path(cmd, path)
+        end
+        yarn_path
+      end
+
+      def yarn_command(command)
+        full_command = "#{yarn_path} #{command}"
+        if options[:quiet]
+          system(full_command, out: File::NULL)
+        else
+          system(full_command)
+        end
+      end
+
+      def run_yarn
```

Comment:
This method should be the first defined and the others be defined after this one.

```
def run_yean
def package_json_exist?
def yarn_path
def yarn_command
```